ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala

```diff
@@ -243,6 +262,86 @@ object KeyValueConsumption {
     )
   }
 
+  private[kvutils] def outOfTimeBoundsEntryToUpdate(
+      recordTime: Timestamp,
+      outOfTimeBoundsEntry: DamlOutOfTimeBoundsEntry): Option[Update] = {
+    val wrappedLogEntry = outOfTimeBoundsEntry.getEntry
+    val duplicateUntilMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasDuplicateUntil,
+      outOfTimeBoundsEntry.getDuplicateUntil)
+    val tooEarlyUntilMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasTooEarlyUntil,
+      outOfTimeBoundsEntry.getTooEarlyUntil)
+    val tooLateFromMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasTooLateFrom,
+      outOfTimeBoundsEntry.getTooLateFrom)
+    val deduplicated = duplicateUntilMaybe.exists(recordTime <= _)
+    val tooEarly = tooEarlyUntilMaybe.exists(recordTime <= _)
+    val tooLate = tooLateFromMaybe.exists(recordTime > _)
+    val invalidLedgerTime = tooEarly || tooLate
+
+    wrappedLogEntry.getPayloadCase match {
+      case _ if deduplicated =>
+        // We don't emit updates for deduplicated submissions.
+        None
+
+      case DamlLogEntry.PayloadCase.TRANSACTION_REJECTION_ENTRY if invalidLedgerTime =>
+        val transactionRejectionEntry = wrappedLogEntry.getTransactionRejectionEntry
+        val reason = (tooEarlyUntilMaybe, tooLateFromMaybe) match {
+          case (Some(lowerBound), Some(upperBound)) =>
+            s"Record time $recordTime outside of range [$lowerBound, $upperBound]"
+          case _ =>
+            "Ledger time outside of valid range"
+        }
+        val rejectionReason = RejectionReason.InvalidLedgerTime(reason)
+        Some(
+          Update.CommandRejected(
+            recordTime = recordTime,
+            submitterInfo = parseSubmitterInfo(transactionRejectionEntry.getSubmitterInfo),
+            reason = rejectionReason
+          )
+        )
+
+      case DamlLogEntry.PayloadCase.CONFIGURATION_REJECTION_ENTRY if invalidLedgerTime =>
+        val configurationRejectionEntry = wrappedLogEntry.getConfigurationRejectionEntry
+        val reason = tooLateFromMaybe
+          .map { maximumRecordTime =>
+            s"Configuration change timed out: $maximumRecordTime < $recordTime"
+          }
+          .getOrElse("Configuration change timed out")
+        Some(
+          Update.ConfigurationChangeRejected(
+            recordTime,
+            SubmissionId.assertFromString(configurationRejectionEntry.getSubmissionId),
+            ParticipantId.assertFromString(configurationRejectionEntry.getParticipantId),
+            Configuration.decode(configurationRejectionEntry.getConfiguration).right.get,
+            reason
+          )
+        )
+
+      case DamlLogEntry.PayloadCase.TRANSACTION_REJECTION_ENTRY |
+          DamlLogEntry.PayloadCase.PACKAGE_UPLOAD_REJECTION_ENTRY |
+          DamlLogEntry.PayloadCase.CONFIGURATION_REJECTION_ENTRY |
+          DamlLogEntry.PayloadCase.PARTY_ALLOCATION_REJECTION_ENTRY =>
+        None
```

Comment:
I wonder whether we should log at INFO or DEBUG here that we've dropped the out-of-time-bounds rejection. I don't expect that this case should happen, so I suggest to at least record in the log that this has happened.

Reply:
Good point -- added at `INFO` level.

Reply:
Shouldn't we rather fail if we don't expect this case to be valid (i.e., seeing it would mean a bug)?

Reply:
It'll have the same effect, i.e., the ledger API server won't stop and you'll have something in the log. However, explicitly logging something will probably yield easier to consume information than an exception caught somewhere.

Reply:
I meant I'd rather crash if we see unexpected behavior, so that we can't possibly overlook it.

Reply:
In order to avoid overlooking such an issue you just need to set up the right monitoring. I.e., in my view the difference between 'crashing' and logging that we've dropped the out-of-time-bounds rejection is that in case of the previous the ledger API server will be restarted automagically whereas in the latter case it doesn't need to be restarted (you'll have the same information in the logs).

Reply:
Yes, if we only catch a bug in production (or any stage with auto-restart) then we need the right monitoring to raise some big red flag, that's why I wonder if a crash (or an exception log) is more amenable to that than an `INFO` log. Maybe raising to `ERROR` at least?

Reply:
Raised to `ERROR`.